src/Microsoft.Identity.Abstractions/ApplicationOptions/CredentialDescription.cs

```diff
@@ -63,17 +67,80 @@ public CredentialDescription(CredentialDescription other)
+                {
+                    switch (SourceType)
+                    {
+                        case CredentialSource.Certificate:
+                            if (Certificate != null)
+                            {
+                                _cachedId = $"Certificate={Certificate.Thumbprint}";
+                            }
+                            break;
+                        case CredentialSource.KeyVault:
+                            _cachedId = $"CertificateFromKeyVault={KeyVaultUrl}/{KeyVaultCertificateName}";
+                            break;
+                        case CredentialSource.Base64Encoded:
+                            _cachedId = $"CertificateFromBase64Encoded={GenerateHash(Base64EncodedValue)}";
+                            break;
+                        case CredentialSource.Path:
+                            _cachedId = $"CertificateFromPath={CertificateDiskPath}";
+                            break;
+                        case CredentialSource.StoreWithThumbprint:
+                            _cachedId = $"CertificateStoreWithThumbprint={CertificateStorePath}/{CertificateThumbprint}";
+                            break;
+                        case CredentialSource.StoreWithDistinguishedName:
+                            _cachedId = $"CertificateStoreWithDistinguishedName={CertificateStorePath}/{CertificateDistinguishedName}";
+                            break;
+                        case CredentialSource.ClientSecret:
+                            _cachedId = $"ClientSecret={GenerateHash(ClientSecret)}";
+                            break;
+                        case CredentialSource.SignedAssertionFromManagedIdentity:
+                            _cachedId = $"SignedAssertionFromManagedIdentity={ManagedIdentityClientId}";
+                            break;
+                        case CredentialSource.SignedAssertionFilePath:
+                            _cachedId = $"SignedAssertionFilePath={SignedAssertionFileDiskPath}";
+                            break;
+                        case CredentialSource.SignedAssertionFromVault:
+                            _cachedId = $"SignedAssertionFromVault={KeyVaultCertificateName}";
+                            break;
+                        case CredentialSource.AutoDecryptKeys:
+                            _cachedId = $"AutoDecryptKeys";
+                            break;
+                        case CredentialSource.CustomSignedAssertion:
+                            string parameterNames = string.Join(",", (IEnumerable<string>?)CustomSignedAssertionProviderData?.Keys ?? []);
+                            _cachedId = $"CustomSignedAssertion={CustomSignedAssertionProviderName}({parameterNames})";
+                            break;
+                        default:
+                            throw new ArgumentException("Unknown credential source type");
+                    }
+                }
+                return _cachedId!;
```

Comment:
This could be null if the source type is cert, and the cert is null.

Reply:
This gets used as a cache key in IdWeb, so this would throw if null and used there. Would be better to throw here if we can't create the id.

Reply:
I'l make it non-null. as I can see cases where the cert would be set afterwards (some customers want to do that)